File: app/src/main/java/com/appchamp/wordchunks/ui/levels/LevelsActivity.java
package com.appchamp.wordchunks.ui.levels;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.appchamp.wordchunks.R;
import com.appchamp.wordchunks.data.LevelsRealmHelper;
import com.appchamp.wordchunks.data.PacksRealmHelper;
import com.appchamp.wordchunks.models.realm.Level;
import com.appchamp.wordchunks.ui.game.GameActivity;
import com.appchamp.wordchunks.ui.packs.PacksActivity;

import java.util.List;

import io.realm.Realm;
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

import static com.appchamp.wordchunks.util.Constants.EXTRA_LEVEL_ID;
import static com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID;


public class LevelsActivity extends AppCompatActivity {

    private Realm realm;
    private RecyclerView rvLevels;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.act_packs_levels);
        LinearLayout llPacksLevels = (LinearLayout) findViewById(R.id.llPacksLevels);
        llPacksLevels.setBackgroundResource(R.drawable.gradient_levels);
        TextView tvTitle = (TextView) findViewById(R.id.tvTitle);
        tvTitle.setText(getString(R.string.title_select_level));
        rvLevels = (RecyclerView) findViewById(R.id.recyclerView);


        realm = Realm.getDefaultInstance();

        // Passing packs via ids
        String packId = getIntent().getStringExtra(EXTRA_PACK_ID);
        if (packId != null) {
            List<Level> levels = LevelsRealmHelper.INSTANCE.findLevelsByPackId(realm, packId);

            final int packColor = Color.parseColor(PacksRealmHelper.INSTANCE
                    .findFirstPackById(realm, levels.get(0).getPackId()).getColor());

            initLevelsAdapter(levels, packColor);
        }
        OverScrollDecoratorHelper.setUpOverScroll(
                rvLevels, OverScrollDecoratorHelper.ORIENTATION_VERTICAL);
    }

    private void initLevelsAdapter(List<Level> levels, int packColor) {
        LevelsAdapter adapter = new LevelsAdapter(levels, packColor);
        rvLevels.setAdapter(adapter);
        rvLevels.setLayoutManager(new LinearLayoutManager(LevelsActivity.this));
        rvLevels.setHasFixedSize(true);

        adapter.setOnItemClickListener((view, position) -> {
            Level clickedLevel = adapter.getItem(position);
            startGameActivity(clickedLevel.getId());
        });
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        startPacksActivity();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close(); // Remember to close Realm when done.
    }

    public void onBackArrowClick(View v) {
        super.onBackPressed();
        startPacksActivity();
    }

    /**
     * Up navigation. Navigates from LevelsActivity to GameActivity passing Level id by the Intent.
     *
     * @param levelId level id to be passed by the Intent.
     */
    private void startGameActivity(String levelId) {
        Intent intent = new Intent(LevelsActivity.this, GameActivity.class);
        intent.putExtra(EXTRA_LEVEL_ID, levelId);
        startActivity(intent);
        finish();
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left);
    }

    /**
     * Back navigation. Navigates from LevelsActivity to PacksActivity.
     */
    private void startPacksActivity() {
        Intent intent = new Intent(LevelsActivity.this, PacksActivity.class);
        realm.executeTransaction(bgRealm -> {
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
            finish();
            overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
        });
    }
}

File: app/src/main/java/com/appchamp/wordchunks/ui/levels/LevelsActivity.kt
/*
 * Copyright 2017 Julia Kozhukhovskaya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.appchamp.wordchunks.ui.levels

import android.arch.lifecycle.Observer
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import com.appchamp.wordchunks.R
import com.appchamp.wordchunks.models.realm.Level
import com.appchamp.wordchunks.ui.BaseActivity
import com.appchamp.wordchunks.ui.game.GameActivity
import com.appchamp.wordchunks.ui.packs.PacksActivity
import com.appchamp.wordchunks.ui.packs.PacksLevelsAdapter
import com.appchamp.wordchunks.util.Constants.EXTRA_LEVEL_ID
import com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID
import com.franmontiel.localechanger.LocaleChanger
import kotlinx.android.synthetic.main.act_packs_levels.*
import kotlinx.android.synthetic.main.titlebar.*
import me.everything.android.ui.overscroll.OverScrollDecoratorHelper
import org.jetbrains.anko.clearTop
import org.jetbrains.anko.intentFor


class LevelsActivity : BaseActivity<LevelsViewModel>() {

    override val viewModelClass = LevelsViewModel::class.java

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.act_packs_levels)

//        llPacksLevels.setBackgroundResource(R.drawable.gradient_levels)
        tvTitle.text = getString(R.string.title_select_level)
        rvList.layoutManager = LinearLayoutManager(this)
        rvList.setHasFixedSize(true)

        // Sets smooth scroll effect.
        OverScrollDecoratorHelper.setUpOverScroll(
                rvList, OverScrollDecoratorHelper.ORIENTATION_VERTICAL)

        btnBack.setOnClickListener { onBackPressed() }

        subscribeUi()
    }

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(LocaleChanger.configureBaseContext(newBase))
    }

    /**
     * Navigates back to PacksActivity.
     */
    override fun onBackPressed() {
        super.onBackPressed()
        backToPacksActivity()
    }

    private fun subscribeUi() {
        // Getting pack id by the Intent.
        val packId = requireNotNull(intent.getStringExtra(EXTRA_PACK_ID),
                { "Activity parameter 'EXTRA_PACK_ID' is missing" })
        // Observe updates to the LiveData levels.
        viewModel
                .getLiveLevels(packId)
                .observe(this, Observer {
                    // update UI
                    val adapter = PacksLevelsAdapter<Level> {
                        // Navigates up to GameActivity passing levelId in the Intent.
                        startGameActivity(it.id)
                    }
                    it?.toList()?.let { levels -> adapter.updateItems(levels) }
                    rvList.adapter = adapter
                })

        val packColor = viewModel.getPackColor() ?: "#7bda7a"
        rlTitlebar.setBackgroundColor(Color.parseColor(packColor))  // get pack color
        // Scrolling RecyclerView to the last "current", or "solved" level item.
        // indexOfLast gets last index, or -1 if the list does not contain that item.
        rvList.smoothScrollToPosition(viewModel.getLastLevelPos())
    }

    private fun startGameActivity(levelId: String) {
        startActivity(intentFor<GameActivity>(EXTRA_LEVEL_ID to levelId).clearTop())
        finish()
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left)
    }

    /**
     * Navigates back to PacksActivity.
     */
    private fun backToPacksActivity() {
        startActivity(intentFor<PacksActivity>().clearTop())
        finish()
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right)
    }
}